docs/sdk/instructlab-sdk.md

```diff
@@ -0,0 +1,109 @@
+4. Define Contracts loose enough that functionality can be expanded as more advanced features are released.
+5. Document SDK usage in upcoming InstructLab releases.
+
+## Non-Goals
+
+1. Exposing all library functionality immediately
+2. Replacing CLI
+3. Shipping an SDK that is generally available as opposed to v1alpha1 or v1beta1.
+
+## Design
+
+### Versioning
+
+The SDK would start at version v1alpha1 such that it can change/break at any time for the first few iterations as libraries adjust their API surface.
+
+### Structure
+
+This SDK should live in a net new package inside of `instructlab/instructlab` preferably to limit unnecessary imports in a new repository. It could be imported as `instructlab.core...`
+
+The user surface initially should look like this:
+
+```console
+instructlab.core.configuration.init
+instructlab.core.taxonomy.diff
+instructlab.core.data.generate
+instructlab.core.data.list
+instructlab.core.model.train
+instructlab.core.model.evaluate
+instructlab.core.model.chat (one shot request probably)
+instructlab.core.model.serve (vLLM and llama-cpp-python)
+instructlab.core.model.download
+instructlab.core.model.list
+instructlab.core.system.info
+instructlab.core.rag.ingest
+instructlab.core.rag.convert
+```
+
+These initial exposed functions can expand to include any new functionality that is more SDK oriented from the various libraries. For example, if SDG adds something like subset selection, teacher as annotator, data mixing, etc we could expose an `instructlab.core.data.annotate` or `instructlab.core.data.mix` that could be invoked in sequence in a user's script with other parts of the ilab workflow. Some things make _less_ sense to be exposed via a CLI, but still are critical to ensuring users get a good model and properly generated data.
+
+There are certain things that only exist in `ilab` currently and functionality that is going to be moving these such as data ingestion, RAG, etc. Forming an SDK for `instructlab` allows us to capture all of these concerns under one API.
+
+These endpoints in combination with the curated InstructLab Config File will open up these workflows to users and allow InstructLab to be easily incorporated into other projects. Allowing people to run things like data generation, and full fine-tuning via an SDK that pulls in their pre-existing config.yaml but also can be run independently will open new avenues for InstructLab adoption and extensibility.
+
+## Changes to the CLI
+
+The `ilab` CLI will need to adapt to this new structure. Commands like `ilab data generate` should, in terms of code, follow this flow:
+
+1. `src/instructlab/cli/data/generate.py`
+2. `src/instructlab/data/generate.py`
+3. `src/instructlab/process.py`
+4. `src/instructlab/api/data/generate.py`
+
+So generally: cli -> internal handling package -> process management package -> api -> library code, is the flow.
```

Comment:
api -> core? (though perhaps "api" is a better name for a module that is meant to be consumed externally. That said, API may also refer to upcoming REST API initiative, so perhaps it's better to avoid this word either. `public` could `sdk` would be another option with clear intent.) Whatever the final name, these should be unified to use just one across the doc.

Reply:
`core` is my preference, personally